### DISCUSSION TITLE
What board to choose and where to buy in europe ?

### DISCUSSION THREAD

**Discussion by masterleo** [grblHAL/Controllers#8] [General]
Hi I build 6 cnc machines already, just wanted to know where I can buy boards and also what your advice would be about the choice of a board for my "Charlyrobot" retrofit. I don't know where to buy and what is up to date / depreciated in terms of hardware. Please do not give me an "experimental, half working board" sugestion as this is for making a production machine I would use a lot. 

please : 
tell me a cheap board  option (20eur)
tell me a decent baord  (50eur)
tell me the best and most expandable one  (100++ euros)

I want to replace my arduino custom made grbl board so that my motors get faster. 
Please give me a good advice as I never used GRBLHAL

**Comment by masterleo** [comment]
Also, is it a good idea to self source and to build myself ? (I can solder anything and program no worries :) )
I was thinking about jlc pcb...

**Reply by terjeio** [reply]
The BSMCE04U-PP is cheap and has optocoupled I/O. For external stepper drivers.

The Teensy 4.1 offerings are both decent and in the best category. Depending on requirements a Pi Pico based board may be a good choice.

The most expandable is Flexi-HAL?

I do not keep track of the state for each listed board, some might be deprecated and unavailable. Follow the links provided and/or search for more information.

**Comment by gharbarka** [comment]
HI, I am trying to migrate from Mach3 to grblHAL having reprogrammed successfully the BSMCE04U-PP board. According to my scope most input outputs are functioning but there are certain inputs whose function i am as yet unable to comprehend. I am sure the issues are mostly my problem.
I have tried to go through the WIKi's but I still have not been able to understand the functions of certain pins.

Is there a forum where one can ask for clarifications?

Finally thanks for an excellent project.

**Reply by terjeio** [reply]
> Is there a forum where one can ask for clarifications?

GitHub discussions, [this for general questions](https://github.com/grblHAL/core/discussions) or for driver specific questions use driver related discussions.

You may also refer to the [Grbl Wiki](https://github.com/gnea/grbl/wiki) for information.
